plugins/org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/syntaxcoloring/DefaultSemanticHighlightingCalculator.java

```diff
@@ -63,22 +69,26 @@ public void provideHighlightingFor(XtextResource resource, IHighlightedPositionA
 	 * @param acceptor
 	 *            the acceptor. Is never <code>null</code>.
 	 */
-	protected void doProvideHighlightingFor(XtextResource resource, IHighlightedPositionAcceptor acceptor) {
-		searchAndHighlightElements(resource, acceptor);
+	protected void doProvideHighlightingFor(XtextResource resource, IHighlightedPositionAcceptor acceptor,
+			CancelIndicator cancelIndicator) {
+		searchAndHighlightElements(resource, acceptor, cancelIndicator);
 		highlightTasks(resource, acceptor);
 	}
 
-	protected void searchAndHighlightElements(XtextResource resource, IHighlightedPositionAcceptor acceptor) {
+	protected void searchAndHighlightElements(XtextResource resource, IHighlightedPositionAcceptor acceptor,
+			CancelIndicator cancelIndicator) {
 		IParseResult parseResult = resource.getParseResult();
 		if (parseResult == null)
 			throw new IllegalStateException("resource#parseResult may not be null");
 		EObject element = parseResult.getRootASTElement();
-		highlightElementRecursively(element, acceptor);
+		highlightElementRecursively(element, acceptor, cancelIndicator);
 	}
 
-	protected void highlightElementRecursively(EObject element, IHighlightedPositionAcceptor acceptor) {
+	protected void highlightElementRecursively(EObject element, IHighlightedPositionAcceptor acceptor,
+			CancelIndicator cancelIndicator) {
 		TreeIterator<EObject> iterator = EcoreUtil2.eAll(element);
 		while (iterator.hasNext()) {
+			operationCanceledManager.checkCanceled(cancelIndicator);
```

Comment:
This appears to be very fine-grained. Maybe we could simply check on each acceptor.accept call by default? Or every 1000 steps?